File: ios/BFExplorer/Plaoc/IPCCore/IPC/IpcResponse.swift
//
//  IpcResponse.swift
//  IPC
//
//  Created by ui03 on 2023/2/13.
//

import UIKit
import Vapor


class IpcResponse {

    var req_id: Int = 0
    var type: IPC_DATA_TYPE = .RESPONSE
    private var headers: IpcHeaders!
    private var statusCode: Int = 0
    private var body: IpcBody?
    private var ipc: Ipc?
    
    required init() {
        
    }
    
    init(req_id: Int, statusCode: Int, headers: IpcHeaders, body: IpcBody, ipc: Ipc) {
        
        self.headers = headers
        self.req_id = req_id
        self.statusCode = statusCode
        self.body = body
        self.ipc = ipc
        
        if let bodySender = body as? IpcBodySender {
            IpcBodySender().usableByIpc(ipc: ipc, ipcBody: bodySender)
        }
    }
    
    lazy var ipcResMessage: IpcResMessage = {
        let message = IpcResMessage(req_id: self.req_id, statusCode: self.statusCode, headers: self.headers, metaBody: self.body?.metaBody)
        return message
    }()
    
    func getIpcHeaders() -> [String:String] {
        return self.headers.headerDict
    }
    
    func toResponse() -> Response {
        
        var headers = HTTPHeaders()
        for (key, value) in self.headers.headerDict {
            headers.add(name: key, value: value)
        }
     
        var responseBody = Response.Body()
        if let content = body?.raw as? String {
            responseBody = Response.Body.init(string: content)
        } else if let tytes = body?.raw as? [UInt8] {
            let data = Data(bytes: tytes, count: tytes.count)
            responseBody = Response.Body.init(data: data)
        } else if let stream = body?.raw as? InputStream {
            let data = IpcResponse.fetchStreamData(stream: stream)
            responseBody = Response.Body.init(data: data)
        }
        
        let response = Response(status: HTTPResponseStatus(statusCode: self.statusCode), headers: headers, body: responseBody)
        return response
    }
    
    static func fetchStreamData(stream: InputStream) -> Data {
        
        stream.open()
        defer {
            stream.close()
        }

        let bufferSize = 1024
        let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
        defer {
            buffer.deallocate()
        }
        
        var result: Data = Data()
        while stream.hasBytesAvailable {
            let length = stream.read(buffer, maxLength: bufferSize)
            let data = Data(bytes: buffer, count: length)
            result.append(data)
        }
        return result
    }
    
    static func fromResponse(req_id: Int, response: Response, ipc: Ipc) -> IpcResponse? {
        
        guard response.body.data != nil else { return nil }
        
        if response.body.count == -1 {
            let stream = InputStream(data: response.body.data!)
            return IpcResponse(req_id: req_id, statusCode: Int(response.status.code), headers: IpcHeaders(content: response.headers.description), body: IpcBodySender(raw: stream, ipc: ipc), ipc: ipc)
        } else if response.body.count > 0 {
            return IpcResponse(req_id: req_id, statusCode: Int(response.status.code), headers: IpcHeaders(content: response.headers.description), body: IpcBodySender(raw: [UInt8](response.body.data!), ipc: ipc), ipc: ipc)
        } else {
            return IpcResponse(req_id: req_id, statusCode: Int(response.status.code), headers: IpcHeaders(content: response.headers.description), body: IpcBodySender(raw: "", ipc: ipc), ipc: ipc)
        }
    }
    
    static func fromJson(req_id: Int,statusCode: Int = 200,jsonable: Any,headers: IpcHeaders = IpcHeaders(),ipc: Ipc)  -> IpcResponse {
        
        headers.set(key: "Content-Type", value: "application/json")
        let ipcResponse = self.fromText(req_id: req_id, statusCode: statusCode, text: ChangeTools.tempAnyToString(value:jsonable) ?? "", headers: headers, ipc: ipc)
        return ipcResponse
    }
    
    static func fromText(req_id: Int,statusCode: Int = 200,text: String,headers: IpcHeaders = IpcHeaders(),ipc: Ipc)  -> IpcResponse {
        
        headers.set(key: "Content-Type", value: "text/plain")
        return IpcResponse(req_id: req_id, statusCode: statusCode, headers: headers, body: IpcBodySender(raw: text, ipc: ipc), ipc: ipc)
    }
    
    static func fromBinary(req_id: Int,statusCode: Int = 200,binary: [UInt8],headers: IpcHeaders,ipc: Ipc)  -> IpcResponse {
        
        headers.set(key: "Content-Type", value: "application/octet-stream")
        headers.set(key: "Content-Length", value: "\(binary.count)")

        return IpcResponse(req_id: req_id, statusCode: statusCode, headers: headers, body: IpcBodySender(raw: binary, ipc: ipc), ipc: ipc)
    }
    
    static func fromStream(req_id: Int,statusCode: Int = 200,stream: InputStream,headers: IpcHeaders = IpcHeaders(),ipc: Ipc) -> IpcResponse {
        
        headers.set(key: "Content-Type", value: "application/octet-stream")
        return IpcResponse(req_id: req_id, statusCode: statusCode, headers: headers, body: IpcBodySender(raw: stream, ipc: ipc), ipc: ipc)
    }
}

extension IpcResponse: IpcMessage {}



struct IpcResMessage: IpcMessage {
    
    var type: IPC_DATA_TYPE = .REQUEST
    var req_id: Int = 0
    var statusCode: Int = 0
    var headers: IpcHeaders = IpcHeaders()
    var metaBody: MetaBody?
    
    init() {
        
    }
    
    init(req_id: Int, statusCode: Int, headers: IpcHeaders, metaBody: MetaBody?) {
        self.req_id = req_id
        self.statusCode = statusCode
        self.headers = headers
        self.metaBody = metaBody
    }
}

File: next/ts/src/core/ipc/IpcResponse.ts
import { once } from "lodash";
import { $Binary, binaryToU8a } from "../../helper/binaryHelper.ts";
import { IpcMessage, IPC_MESSAGE_TYPE } from "./const.ts";
import type { Ipc } from "./ipc.ts";
import type { IpcBody } from "./IpcBody.ts";
import { IpcBodySender, setStreamId } from "./IpcBodySender.ts";
import { IpcHeaders } from "./IpcHeaders.ts";
import type { MetaBody } from "./MetaBody.ts";

export class IpcResponse extends IpcMessage<IPC_MESSAGE_TYPE.RESPONSE> {
  constructor(
    readonly req_id: number,
    readonly statusCode: number,
    readonly headers: IpcHeaders,
    readonly body: IpcBody,
    readonly ipc: Ipc
  ) {
    super(IPC_MESSAGE_TYPE.RESPONSE);
    if (body instanceof IpcBodySender) {
      IpcBodySender.$usableByIpc(ipc, body);
    }
  }

  #ipcHeaders?: IpcHeaders;
  get ipcHeaders() {
    return (this.#ipcHeaders ??= new IpcHeaders(this.headers));
  }

  toResponse(url?: string) {
    const body = this.body.raw;
    if (body instanceof Uint8Array) {
      this.headers.init("Content-Length", body.length + "");
    }
    const response = new Response(body, {
      headers: this.headers,
      status: this.statusCode,
    });
    if (url) {
      Object.defineProperty(response, "url", {
        value: url,
        enumerable: true,
        configurable: true,
        writable: false,
      });
    }
    return response;
  }

  /** 将 response 对象进行转码变成 ipcResponse */
  static async fromResponse(
    req_id: number,
    response: Response,
    ipc: Ipc,
    /// 如果有 content-length，说明大小是明确的，不要走流，直接传输就好，减少 IPC 的触发次数. TODO 需要注意大小是否过大，过大的话还是要分片传输。不过这种大二进制情况下一般是请求文件，应该直接使用句柄转发
    asBinary = false // response.headers.get("content-length") !== null
  ) {
    if (response.bodyUsed) {
      throw new Error("body used");
    }
    let ipcBody: IpcBody;
    if (asBinary || response.body == undefined) {
      ipcBody = IpcBodySender.from(
        binaryToU8a(await response.arrayBuffer()),
        ipc
      );
    } else {
      setStreamId(response.body, response.url);
      ipcBody = IpcBodySender.from(response.body, ipc);
    }

    return new IpcResponse(
      req_id,
      response.status,
      new IpcHeaders(response.headers),
      ipcBody,
      ipc
    );
  }
  static fromJson(
    req_id: number,
    statusCode: number,
    headers = new IpcHeaders(),
    jsonable: unknown,
    ipc: Ipc
  ) {
    headers.init("Content-Type", "application/json");
    return this.fromText(
      req_id,
      statusCode,
      headers,
      JSON.stringify(jsonable),
      ipc
    );
  }
  static fromText(
    req_id: number,
    statusCode: number,
    headers = new IpcHeaders(),
    text: string,
    ipc: Ipc
  ) {
    headers.init("Content-Type", "text/plain");
    // 这里 content-length 默认不写，因为这是要算二进制的长度，我们这里只有在字符串的长度，不是一个东西
    return new IpcResponse(
      req_id,
      statusCode,
      headers,
      IpcBodySender.from(text, ipc),
      ipc
    );
  }
  static fromBinary(
    req_id: number,
    statusCode: number,
    headers = new IpcHeaders(),
    binary: $Binary,
    ipc: Ipc
  ) {
    headers.init("Content-Type", "application/octet-stream");
    headers.init("Content-Length", binary.byteLength + "");
    return new IpcResponse(
      req_id,
      statusCode,
      headers,
      IpcBodySender.from(binaryToU8a(binary), ipc),
      ipc
    );
  }
  static fromStream(
    req_id: number,
    statusCode: number,
    headers = new IpcHeaders(),
    stream: ReadableStream<Uint8Array>,
    ipc: Ipc
  ) {
    headers.init("Content-Type", "application/octet-stream");
    const ipcResponse = new IpcResponse(
      req_id,
      statusCode,
      headers,
      IpcBodySender.from(stream, ipc),
      ipc
    );
    return ipcResponse;
  }

  readonly ipcResMessage = once(
    () =>
      new IpcResMessage(
        this.req_id,
        this.statusCode,
        this.headers.toJSON(),
        this.body.metaBody
      )
  );
  toJSON() {
    return this.ipcResMessage();
  }
}

export class IpcResMessage extends IpcMessage<IPC_MESSAGE_TYPE.RESPONSE> {
  constructor(
    readonly req_id: number,
    readonly statusCode: number,
    readonly headers: Record<string, string>,
    readonly metaBody: MetaBody
  ) {
    super(IPC_MESSAGE_TYPE.RESPONSE);
  }
}
